### PR TITLE
feat: add warning about default filter override

### PR DESCRIPTION
**Pull request by magnew**
Closes: #11642 

### Description:

Adds a warning modal when editing a dashboard with filters changed:
Screenshot 2024-11-12 at 17 34 40

If the user clicks yes, we clear the filters and go into edit mode. If they click cancel, cancel editing. 

To repro: either update a saved filter on a dashboard, or add a temp fitler. Then try to edit. 

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/pages/Dashboard.tsx [hunk 1]
```diff
@@ -732,6 +757,13 @@ const Dashboard: FC = () => {
                         onConfirm={duplicateModalHandlers.close}
                     />
                 )}
+                {isFilterWarningModalOpen && (
+                    <DashboardFiltersWarningModal
+                        onConfirm={handleEnterEditMode}
+                        onClose={() => setIsFilterWarningModalOpen(false)}
+                        opened={isFilterWarningModalOpen}
+                    />
+                )}
```
Comment (joaoviana) [inline packages/frontend/src/pages/Dashboard.tsx:766]:
```suggestion

                    <DashboardFiltersWarningModal
                        onConfirm={handleEnterEditMode}
                        onClose={() => setIsFilterWarningModalOpen(false)}
                        opened={isFilterWarningModalOpen}
                    />
```

NP: this is sufficient

File: packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx [hunk 2]
```diff
@@ -84,9 +96,14 @@ export const useSavedDashboardFiltersOverrides = () => {
         dispatch({ type: REMOVE_SAVED_FILTER_OVERRIDE, payload: item });
     };
 
+    const resetSavedFilterOverrides = () => {
+        dispatch({ type: RESET_SAVED_FILTER_OVERRIDES, payload: null });
```
Comment (joaoviana) [inline packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx:100]: do we need `payload: null`

File: packages/frontend/src/components/common/modal/DashboardFiltersWarningModal.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    onConfirm,
+    ...modalProps
+}) => {
+    return (
+        <Modal withCloseButton={false} {...modalProps}>
+            <Stack pt="sm">
+                <Text>
+                    Your current filter values do not match the defaults for
+                    this dashboard. Do you want to continue to edit and lose the
+                    current values?
```
Comment (joaoviana) [inline packages/frontend/src/components/common/modal/DashboardFiltersWarningModal.tsx:25]:
```suggestion
                    Your current filter values do not match the defaults for
                    this dashboard. Do you want to discard your changes and continue?
```

File: packages/frontend/src/components/common/Dashboard/DashboardHeader.tsx [hunk 4]
```diff
@@ -374,11 +375,7 @@ const DashboardHeader = ({
                         >
                             <ActionIcon
                                 variant="default"
-                                onClick={() => {
-                                    history.replace(
-                                        `/projects/${projectUuid}/dashboards/${dashboardUuid}/edit`,
-                                    );
-                                }}
+                                onClick={onEditClicked}
```
Comment (joaoviana) [inline packages/frontend/src/components/common/Dashboard/DashboardHeader.tsx:378]: so the `resetDashboardFilters` will do this `history` replace for us?

File: packages/frontend/src/pages/Dashboard.tsx (hunk 1 above)
Comment (magnew) [reply packages/frontend/src/pages/Dashboard.tsx:766]: Will do. We often follow that existing pattern to intentionally mount and unmount the modal, but this one has no state so it doesn't matter.

File: packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx (hunk 2 above)
Comment (magnew) [reply packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx:100]: It avoids some more convoluted type handling. The action wants a payload. This seems simple and explicit but if you have a pattern you prefer, I can change it!

File: packages/frontend/src/components/common/Dashboard/DashboardHeader.tsx (hunk 4 above)
Comment (magnew) [reply packages/frontend/src/components/common/Dashboard/DashboardHeader.tsx:378]: `handleEnterEditMode` calls `resetDashboardFilters`, then does the redirect

File: packages/frontend/src/components/common/modal/DashboardFiltersWarningModal.tsx (hunk 3 above)
Comment (magnew) [reply packages/frontend/src/components/common/modal/DashboardFiltersWarningModal.tsx:25]: I just took this from the ticket, but I like your version better.

File: packages/frontend/src/pages/Dashboard.tsx (hunk 1 above)
Comment (magnew) [reply packages/frontend/src/pages/Dashboard.tsx:766]: Actually I'm gonna keep this for consistency with all of the other modals on this page, if that's cool. There are a lot of them. And un-mounting the unopened ones seems ok.

Comment (joaoviana) [reply packages/frontend/src/pages/Dashboard.tsx:766]: makes sense!

File: packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx (hunk 2 above)
Comment (joaoviana) [reply packages/frontend/src/hooks/useSavedDashboardFiltersOverrides.tsx:100]: understood!

**Review by joaoviana — APPROVED** [review]

**Pull request merged**